ed25519.go

```diff
@@ -63,10 +63,62 @@ const (
 
 	// SeedSize is the size, in bytes, of private key seeds. These are the private key representations used by RFC 8032.
 	SeedSize = 32
+
+	// ContextMaxSize is the maximum allowed context length for Ed25519ctx.
+	ContextMaxSize = 255
 )
 
 var _ crypto.Signer = (PrivateKey)(nil)
 
+// Options can be used with PrivateKey.Sign or VerifyWithOptions
+// to select Ed25519 variants.
+type Options struct {
+	// Hash can be crypto.Hash(0) for Ed25519/Ed25519ctx, or crypto.SHA512
+	// for Ed25519ph.
+	Hash crypto.Hash
+
+	// Context is an optional domain separation context for Ed2551ph and
+	// Ed25519ctx. It must be less than or equal to ContextMaxSize
+	// in length.
+	Context string
+}
+
+// HashFunc returns an identifier for the hash function used to produce
+// the message pased to Signer.Sign. For the Ed25519 family this must
+// be crypto.Hash(0) for Ed25519/Ed25519ctx, or crypto.SHA512 for
+// Ed25519ph.
+func (opt *Options) HashFunc() crypto.Hash {
+	return opt.Hash
+}
+
+func unwrapOptions(message []byte, opts crypto.SignerOpts) (dom2Flag, []byte, error) {
+	var (
+		context []byte
+		f       dom2Flag = fPure
+	)
+	if o, ok := opts.(*Options); ok {
+		context = []byte(o.Context)
+		if l := len(context); l > ContextMaxSize {
+			return f, nil, errors.New("ed25519: bad context length: " + strconv.Itoa(l))
+		}
+
+		f = fCtx
```

Comment:
So as soon as options are specified, this will switch to at least Ed25519ctx. So no way to have options and pure Ed25519? I know that it doesn't make sense to specify options in this case (as they would be empty anyway), but IMO the API would be cleaner if a default-constructed SignerOpts behaved the same as nil SignerOpts.

Reply:
Changed it slightly, left a comment.